staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go

```diff
@@ -145,9 +149,44 @@ func (s *SelectionPredicate) Empty() bool {
 // For any index defined by IndexFields, if a matcher can match only (a subset)
 // of objects that return <value> for a given index, a pair (<index name>, <value>)
 // wil be returned.
-func (s *SelectionPredicate) MatcherIndex() []MatchValue {
+func (s *SelectionPredicate) MatcherIndex(ctx context.Context) []MatchValue {
```

Comment:
I'm not a fan of the code below - I think sth like that would be easier:
```
for _, field := range s.IndexFields {
	if value, ok := s.Field.RequiresExactMatch(field); ok {
		result = append(result, MatchValue{IndexName: FieldIndex(field), Value: value})
	} else if field == "metadata.namespace" {
		if namespace, isNamespaceScope := isNamespaceScopedRequest(ctx); isNamespaceScope {
			result = append(result, MatchValue{IndexName: FieldIndex(field), Value: namespace})
		}
	}
	...
```

namespace.name is somewhat special anyway, so I think that making this explicit would make it much easier to follow

Reply:
Updated.

Reply:
I very consciously wrote what is above and what you did is not that.
There are couple reasons I did what I did being:
(a) the other pod index is `spec.nodeName` and in many cases that is much more limiting index.
  So if such fieldSelector is specified we actually want to prioritize for it [and that is set as the first one in `IndexFields`
(b) let's not even try to look at request info if we can use `nodeName` index
(c) if FieldSelector itself is setting `metadata.namespace` let's prefer this one

Reply:
Thanks for your detail explanation.  I consider your code, but there is one case it doesn't handle correctly. i.e. `/api/v1/namespaces/default/pods?fieldSelector=metadata.namespace=kube-system`. The above code will try to get pods from `kube-system` namespace and then filter data by the namespace `kube-system` where it's not correct.
For the `spec.nodeName`, you're right and I miss the situation.  I update the codes.